### ISSUE TITLE
`@eval` into a module breaks subtyping relationships

### ISSUE THREAD

**Issue by Seelengrab** [JuliaLang/julia#50354]
See also https://github.com/JuliaDocs/Documenter.jl/issues/2160

```
julia> const m = Module()
Main.anonymous

julia> @eval m begin
           abstract type Foo end
           struct Bar <: Foo end
       end

julia> supertype(m.Bar) === m.Foo
true

julia> subtypes(m.Foo)
Any[]
```

`subtypes` should  obviously contain `m.Bar` as well. Whether the module is anonymous or not doesn't matter.

**Comment by nsajko** [comment]
Subtyping itself (`<:`) works fine, it's just `subtypes` that's broken, and only for anonymous modules, it seems.

**Comment by Seelengrab** [comment]
If `A <: B`, then `A in subtypes(B)` must hold, that is an invariant of subtyping. If it's broken, subtyping is broken; there is no other way to retrieve A if you only have access to B.

**Comment by nsajko** [comment]
We're arguing about semantics I guess, which is a bit pointless, but anyway `subtypes` is just a stdlib helper for interactive use, while subtyping is an essential part of Julia. As far as I see `subtypes` is the only thing that doesn't work, so I'd adjust the title if it were in my power.

**Comment by fingolfin** [comment]
> Whether the module is anonymous or not doesn't matter.

What makes you say that? The problem goes away for me if the module is named:
```
julia> module Quux
       end
Main.Quux

julia> @eval Quux begin
           abstract type Foo end
           struct Bar <: Foo end
       end

julia> supertype(Quux.Bar) === Quux.Foo
true

julia> subtypes(Quux.Foo)
1-element Vector{Any}:
 Main.Quux.Bar
```

**Comment by nsajko** [comment]
I think @Seelengrab's point was that whether the module is named *should* not matter, which certainly makes sense.

**Comment by Seelengrab** [comment]
No, I was definitely referring to named modules:

```
julia> const m = Module(:foo)
Main.foo

julia> @eval m begin
           abstract type Foo end
           struct Bar <: Foo end
       end

julia> supertype(m.Bar) === m.Foo
true

julia> subtypes(m.Foo)
Any[]
```

It's just that apparently creating `Module`s by hand, instead of through the parser, is broken/internal.

**Comment by MasonProtter** [comment]
It's not internal, since `Module` is public, documented API. This seems to be a (relatively serious?) bug in reflection

**Comment by Seelengrab** [comment]
You're right, those do seem to be `public`  - I could swear some core dev once mentioned that this is kind of internal 🤔
